packages/web3-eth-abi/src/coders/base/number.ts

```diff
@@ -0,0 +1,176 @@
+		result <<= BigInt(8);
+		// eslint-disable-next-line no-bitwise
+		result |= BigInt(byte);
+	}
+	// If negative, apply two's complement
+	if (isNegative) {
+		// eslint-disable-next-line no-bitwise
+		const mask = (BigInt(1) << (BigInt(8) * BigInt(value.length))) - BigInt(1);
+		// eslint-disable-next-line no-bitwise
+		result = -((~result + BigInt(1)) & mask);
+	}
+
+	return result;
+}
+
+const numberLimits = new Map<string, { min: bigint; max: bigint }>();
+
+// precalculate all the limits
+for (let i = 8; i <= 256; i += 8) {
+	numberLimits.set(`uint${i}`, {
+		min: BigInt(0),
+		max: BigInt(2) ** BigInt(i) - BigInt(1),
+	});
+	numberLimits.set(`int${i}`, {
+		min: -(BigInt(2) ** BigInt(i - 1)),
+		max: BigInt(2) ** BigInt(i - 1) - BigInt(1),
+	});
+}
+
+// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+numberLimits.set(`int`, numberLimits.get('int256')!);
+// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+numberLimits.set(`uint`, numberLimits.get('uint256')!);
+
+export function encodeNumber(param: AbiParameter, input: unknown): EncoderResult {
+	let value;
+	try {
+		value = toBigInt(input);
+	} catch (e) {
+		throw new AbiError('provided input is not number value', {
+			type: param.type,
+			value: input,
+			name: param.name,
+		});
+	}
+	const limit = numberLimits.get(param.type);
+	if (!limit) {
+		throw new AbiError('provided abi contains invalid number datatype', { type: param.type });
+	}
+	if (value < limit.min) {
+		throw new AbiError('provided input is less then minimum for given type', {
+			type: param.type,
+			value: input,
+			name: param.name,
+			minimum: limit.min.toString(),
+		});
+	}
+	if (value > limit.max) {
+		throw new AbiError('provided input is greater then maximum for given type', {
+			type: param.type,
+			value: input,
+			name: param.name,
+			maximum: limit.max.toString(),
+		});
+	}
+	return {
+		dynamic: false,
+		encoded: bigIntToUint8Array(value),
+	};
+}
+
+export function decodeNumber(param: AbiParameter, bytes: Uint8Array): DecoderResult<bigint> {
+	if (bytes.length < WORD_SIZE) {
+		throw new AbiError('Not enough bytes left to decode', { param, bytesLeft: bytes.length });
+	}
+	const boolBytes = bytes.subarray(0, WORD_SIZE);
+	const limit = numberLimits.get(param.type);
+	if (!limit) {
+		throw new AbiError('provided abi contains invalid number datatype', { type: param.type });
+	}
+	const numberResult = uint8ArrayToBigInt(boolBytes);
+
+	if (numberResult < limit.min) {
+		throw new AbiError('decoded value is less then minimum for given type', {
+			type: param.type,
+			value: numberResult,
+			name: param.name,
+			minimum: limit.min.toString(),
+		});
+	}
+	if (numberResult > limit.max) {
+		throw new AbiError('decoded value is greater then maximum for given type', {
+			type: param.type,
+			value: numberResult,
+			name: param.name,
+			maximum: limit.max.toString(),
+		});
+	}
+	return {
+		result: numberResult,
+		encoded: bytes.subarray(WORD_SIZE),
```

Comment:
subarray creates a new view on existing buffer, and changes on new obj will impact original and vice versa. so I think a copy should be made of bytes with required data and sent back from this function.

Reply:
same for other places of code in this PR

Reply:
ok, Its tradeoff between memory+performance vs safety( so data is not corrupted by maintainers misuse of it). 

I'll also lean towards UX as compared to maintainers. btw I like this approach of using uint8arry / buffers instead of padding using high level functions, its great.